### PR TITLE
fix: handle pure ESM robustly

### LINKED ISSUE #427: preset-prettier integration broken after update of remark-preset-prettier to 2.0.0
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

eslint-plugin-mdx 2.0.4

### Link to runnable example

Not runnable, but easily made runnable: https://github.com/enqueue/eslint-mdx-reproducer

### Steps to reproduce

1. Set up a project with eslint-plugin-mdx and remark-preset-prettier version 2.0.0
2. Add "preset-prettier" plugin to projects `remarkrc`
3. Run `eslint .`

(Hopefully) minimal reproducible example over here: https://github.com/enqueue/eslint-mdx-reproducer

Node version is 18.9

### Expected behavior

MDX and MD files are linted according to rules.

### Actual behavior

Error message `0:0  error  Parsing error: Cannot find module 'preset-prettier'`.

### Runtime

Other (please specify in steps to reproduce)

### Package manager

npm v8

### OS

Linux

### Build and bundle tools

_No response_

### Workaround

My current workaround is to pin the dependency of remark-preset-prettier to 1.0.2

### PR DESCRIPTION
**Pull request by JounQin**
close #427

### CODE REVIEW

File: packages/eslint-mdx/src/helpers.ts [hunk 1]
```diff
@@ -90,8 +90,14 @@ export const loadModule = async <T>(modulePath: string): Promise<T> => {
...
         return require(modulePath)
       } catch (err) {
+        const code = (err as { code: string }).code
         /* istanbul ignore if */
-        if ((err as { code: string }).code === 'ERR_REQUIRE_ESM') {
+        if (
+          code === 'ERR_REQUIRE_ESM' ||
+          // A pure ESM could have no `exports.require` and then throw the following error,
+          // related to #427.
+          code === 'ERR_PACKAGE_PATH_NOT_EXPORTED'
```
Comment (wooorm) [inline packages/eslint-mdx/src/helpers.ts:99]: Can‘t you use `import-meta-resolve`? That everything from conditions to import/export maps, etc.

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:99]: @wooorm `import-meta-resolve` is pure ESM only, `eslint-mdx` is commonjs, I don't understand how will it help? Any advice?

Comment (wooorm) [reply packages/eslint-mdx/src/helpers.ts:99]: > `eslint-mdx` is commonjs

I forgot that this project is still CJS.
Still, you could bundle `import-meta-resolve` (`npx esbuild import-meta-resolve --bundle --platform=node --outfile=vendor/import-meta-resolve.js`), or if the wrapping code is async, use a dynamic import expression to get it (`await import('xxx')`).

>  I don't understand how will it help? Any advice?

`import-meta-resolve` is 1500 lines (sloc). That’s a lot of code. It matches how Node imports ESM and CJS. The code here is about 50 lines, so it’s missing some things.

This PR is called “handle pure ESM robustly”, if you want to handle ESM/CJS really robustly, like Node.js does, then I recommend `import-meta-resolve`!

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:99]: This PR handlers pure ESM remark plugins more robustly, not itself, sorry it could be confusing for you.

Comment (wooorm) [reply packages/eslint-mdx/src/helpers.ts:99]: You are adding an exception message for export maps.
Your code here does not handle export maps.
`import-meta-resolve` handles export maps.

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:99]: `loadEsmModule` is using `import()` to load modules.

```ts
// Load the ESM configuration file using the TypeScript dynamic import workaround.
// Once TypeScript provides support for keeping the dynamic import this workaround can be
// changed to a direct dynamic import.
```

I'll change this project by using `node16` resolution, but that would a minor change.

Comment (JounQin) [reply packages/eslint-mdx/src/helpers.ts:99]: @wooorm I merged this PR for not blocking a patch release, I'll work on a minor change to use node16 resolution and native dynamic import later.

**Pull request merged**